### ISSUE TITLE
Assets with dynamic names are not resolved

### ISSUE THREAD

**Issue by tobiasdiez** [nuxt/nuxt.js#14766]

**Comment by Umer-Farooq10** [comment]
For dynamic image use this

`<img :src="`/_nuxt/assets/images/${img.src}`" alt="image"/>`

Note: define your directory properly, my image are saved in **images** folder.

It work for me :)

**Comment by Umer-Farooq10** [comment]
For dynamic image use this

`<img :src="\`/_nuxt/assets/images/${img.src}\`" alt="image"/>`

Note: define your directory properly, my image are saved in **images** folder.

It work for me :)

**Comment by Umer-Farooq10** [comment]
For dynamic image use this

```<img :src="`/_nuxt/assets/images/${img.src}`" alt="image"/>```

Note: define your directory properly, my image are saved in **images** folder.

It work for me :)

**Comment by manniL** [comment]
@Umer-Farooq10 this will break in prod

**Comment by saslavik** [comment]
I can't explain why, but this fn not working
`const getImg =  (img) => {
      return new URL(`./../assets/images/useful/${img}.png`, import.meta.url).href
    }`

and this one is working for me O_o :

`const getImg =  (img) => {
      const link = `./../assets/images/useful/${img}.png`
      return new URL(link, import.meta.url).href
    }`

**Comment by bigyonk** [comment]
> I believe this is vite and @vitejs/plugin-vue behaviour. It's a bit complicated, but you can do this with `import.meta.glob`:
> 
> https://stackblitz.com/edit/github-pq8nym-aoavsh
> 
> ```
> <template>
>   <img :src="images[dynamic_image_name]" alt="Discover Nuxt 3" />
> </template>
> <script setup lang="ts">
> import { filename } from 'pathe/utils';
> 
> const glob = import.meta.glob('~/assets/*.svg', { eager: true });
> const images = Object.fromEntries(
>   Object.entries(glob).map(([key, value]) => [filename(key), value.default])
> );
> 
> const dynamic_image_name = 'zero-config';
> </script>
> ```

@danielroe Can this be accomplished with the Options API?

**Comment by danielroe** [comment]
@netopolit Indeed they will create chunks for all files in that directory. But there's no other way to handle fully dynamic imports, because the bundler _has_ to include everything in chunks because it doesn't know what variable you will pass at runtime.

**Comment by katerlouis** [comment]
As great as Nuxt3 already, hurdles like this where Nuxt2 was simply better, isn't easy to swallow. I also liked to get img dimensions using require. 

@danielroe does the core team have ideas on how to provide this or is this  not a concern for the foreseeable future? 

I suggest mentioning things missing from Nuxt2 to the migration guide, so people are aware of the current limitations:
https://nuxt.com/docs/migration/overview

**Comment by luksak** [comment]
> As great as Nuxt3 already, hurdles like this where Nuxt2 simply was better, are frustrating. I also liked to get img dimensions using require.
> 
> @danielroe does the core team have ideas on how to support these usecases or is this not a concern for the foreseeable future?
> 
> I suggest mentioning things missing from Nuxt2 to the migration guide, so people are aware of the current limitations:
> https://nuxt.com/docs/migration/overview

Determining image dimensions is critical for leveraging good performance with  images, since this is the only way to have lazy loading and prevent layout shifts. Actually we don't need the actual image for that. We just need to get the dimensions. Is there any way to accomplish this?

Do we need a separate issue for this?

**Comment by miixel2** [comment]
Surprisingly that even in official document addressed we can use following syntax but it doesn't work.

```
<template>
  <img src="~/assets/img/nuxt.png" alt="Discover Nuxt 3" />
</template>
```
Ref: https://nuxt.com/docs/getting-started/assets

**Comment by danielroe** [comment]
@miixel2 This works fine for me. If it's not working for you, please raise an issue with a reproduction 🙏

**Comment by miixel2** [comment]
Thanks to alternative solutions by @danielroe and @tobiasdiez 

I ended up with following solution.

1. Create useAssets to composables which has three functions
   - useAssets(path: string) to resolve the image path
   - getImages to get all available images, svg png jpg and jpeg, in assets dir.

```typescript
// composables/useAssets.ts
import { filename } from 'pathe/utils'
export default function useAssets(path: string) {
    let assets
    if (/(\.svg)$/i.exec(path)) assets = import.meta.glob('~/assets/img/**/*.svg', { eager: true })
    else if (/(\.png)$/i.exec(path)) assets = import.meta.glob('~/assets/img/**/*.png', { eager: true })
    else assets = import.meta.glob('~/assets/img/**/*.jpg', { eager: true })
    const fileName = filename(path)
    const images = Object.fromEntries(
        Object.entries(assets).map(([key, value]) => [filename(key), (value as Record<string, any>).default]),
    )
    return images[fileName]
}

export const getImages = () => {
    const assets = import.meta.glob('~/assets/img/**/*.(svg|png|jpg|jpeg)', { eager: true })
    return Object.fromEntries(
        Object.entries(assets).map(([key, value]) => [filename(key), (value as Record<string, any>).default]),
    )
}

```
2. Usage
  - `useAssets` 
       We can input just only image name without any relative path. The function will handle the rest for us.
       For example, the image.png is located in `'assets/img/image.png'`. We can just put `'image.png'` in the `useAssets` function.
```vue
<template>
    <img :src="useAssets('image.png')" alt="Item Top Tab" />
</template>
<script lang="ts" setup>
    import { useAssets } from '~/composables/useAssets'
</script>
```
  - `getImages`
     To get list of the image and use as variable
     Assume that we have `image1.png` and `image2.png` under `'/assets/img'` dir.
```vue
<template>
    <img :src="image1" alt=" image 1" />
    <img :src="image2" alt=" image 2" />
</template>
<script lang="ts" setup>
import { getImages } from '~/composables/useAssets'
const { image1,  image2 } = getImages()
</script>
```

**Comment by miixel2** [comment]
> 

I will try create an issue about this thanks again.

**Comment by Umer-Farooq10** [comment]
> @Umer-Farooq10 this will break in prod

Then what is the right/proper solution?

**Comment by Umer-Farooq10** [comment]
> Thanks @danielroe, this works like a charm! Do you think it makes sense to add the following helper method as a built-in composable?
> 
> ```ts
> function useAsset(path: string): string {
>   const assets = import.meta.glob('~/assets/**/*', {
>     eager: true,
>     import: 'default',
>   })
>   // @ts-expect-error: wrong type info
>   return assets['/assets/' + path]
> }
> 
> //Usage: <img :src="useAsset(dynamic_image_name + '.svg')" alt="Discover Nuxt 3" />
> ```

This work for me. I use it as a Composables. Thank you :)

**Comment by scofield-ua** [comment]
> Thanks @danielroe, this works like a charm! Do you think it makes sense to add the following helper method as a built-in composable?
> 
> ```ts
> function useAsset(path: string): string {
>   const assets = import.meta.glob('~/assets/**/*', {
>     eager: true,
>     import: 'default',
>   })
>   // @ts-expect-error: wrong type info
>   return assets['/assets/' + path]
> }
> 
> //Usage: <img :src="useAsset(dynamic_image_name + '.svg')" alt="Discover Nuxt 3" />
> ```

For client-side render this will add all imported files into `head` tag with prefetch mode, which will force user to load all of the assets. 

Also it will put all import\resolve references into a file which can become big very quickly if you have a lot of assets.

**Comment by fullstuffdeveloper** [comment]
export const useAsset = (path: string): string => {
  const assets = import.meta.glob("~/assets/icons/*", {
    eager: true,
    import: "default"
  });
  // @ts-expect-error: wrong type info
  return assets["/assets/icons/" + path];
};

**Comment by stbit** [comment]
```
const getImgSrc = (() => {
  const assets = import.meta.glob('~/assets/**/*', {
    eager: true,
    as: 'url', // get only url path with hash
  })

  return (path: string): string => {
    return assets['/assets/' + path]
  }
})()

useAsset(`img/${dynamicName}.jpg`)
```

**Comment by szulcus** [comment]
> I created a temporary composable function while waiting for an official solution.
> 
> It uses the above solution with good types as much as possible (I didn't succeed to have the correct `Module` returned by `import.meta.glob` function).
> 
> ```ts
> import { filename } from "pathe/utils";
> import { computed } from "#build/imports";
> 
> interface ImagesComposable {
>   getImageSrc: (fileName: string) => string | undefined;
> }
> 
> function useImages(): ImagesComposable {
>   // TODO : replace the first parameter of the glob function according to your needs, I needed to import only png and jpeg from the images directory.
>   const images = computed(() => import.meta.glob("~/assets/images/*.(png|jpeg)", { eager: true }));
> 
>   const getImageSrc = (fileName: string): string | undefined => {
>     for (const path in images.value) {
>       if (Object.hasOwn(images.value, path)) {
>         // unknown type is required here to change the final type as typescript thinks that images.value[path] is a function, it is not.
>         const image: unknown = images.value[path];
>         const imagePath = (image as { default: string }).default;
>         if (filename(imagePath) === filename(fileName)) {
>           return imagePath;
>         }
>       }
>     }
>     return undefined;
>   };
>   return { getImageSrc };
> }
> 
> export { useImages };
> 
> // USAGE => const src = getImageSrc("test.png")
> ```

Your function unfortunately doesn't distinguish between files with the same name that are nested in different folders. I changed it up a bit, but thanks for the inspiration!
```ts
const useAssets = () => {
	// Images
	const images = computed<Record<string, { default: string }>>(() => import.meta.glob('#nuxt-layer-base/assets/images/**/*.(png|jpeg|svg)', { eager: true }));
	const getImage = (src: string): string | undefined => {
		for (const path in images.value) {
			const image = images.value[path].default;
			if (path.endsWith(`assets/images/${src}`)) return image;
		}
		return undefined;
	};

	return {
		getImage,
	};
};

export default useAssets;
```

**Comment by netopolit** [comment]
@danielroe Is it currently possible to import **a single** image by dynamic name without bundling all the images `import.meta.glob` finds?

Thanks

**Comment by danielroe** [comment]
@netopolit - How would it know that it is only a single image you might need? The point is that it can't know which one, so it has to bundle all of them.

**Comment by netopolit** [comment]
@danielroe Thank you. Wish it was easy like it was with CommonJS require.

**Comment by danielroe** [comment]
Quite so. If you prefer the previous behaviour, you can of course use the Nuxt 3 webpack builder. The difference in behaviour you are referring to is the difference between webpack and vite, not between Nuxt 3 and Nuxt 2.

**Comment by MartinsOnuoha** [comment]
Yeah, you're right, technically, in the end, it's a choice of default/out-of-the-box build/dev tools.
Not to downplay the benefits, of course. It's just devs wanting to avoid an "extra" build setup, however minute.

**Comment by victorgarciaesgi** [comment]
Is there a way from a module to inject build assets? 
I have a module that needs to inject svg icons to be resolved by `vite-svg-loader` in the target app, but couldn't find anything on nuxt or nitro config

**Comment by Werhww** [comment]
Putting "_nuxt" at the start of the image link worked for me. Like this:
<img :src="'_nuxt/assets/' + image">

This also worked for images in public.

**Comment by szulcus** [comment]
Now try to deploy ;)

**Comment by Red-Asuka** [comment]
When there are many images, this approach will significantly increase the initial loading size of JS files. It is urgent to find a way to import only specified resources.

**Comment by IJsLauw** [comment]
I'd like to try out your code, but got stuck somehow.

> ```ts
> import.meta.glob('#nuxt-layer-base/assets/images/**/*.(png|jpeg|svg)', { eager: true }
> ```

This part has me stumped. When I try this all I get is an error like this:

`[plugin:vite:import-glob] Invalid glob: "#nuxt-layer-base/assets/images/**/*.(png|jpeg|svg)" (resolved: "#nuxt-layer-base/assets/images/**/*.(png|jpeg|svg)"). It must start with '/' or './'
.... /composables/assets.ts`

Any thoughts on that @szulcus ?

**Comment by lionel-addvanto** [comment]
is there any update on this? Anything official from the Nuxt team maybe? @danielroe?

**Comment by matias-gmg** [comment]
Also interested in how to address this.

**Comment by matias-gmg** [comment]
I solved it by using the following code:

```
<template>
  <span class="svg-icon-ctr" :class="[name, cls]" v-html="icon"></span>
</template>

<script setup>
const props = defineProps({
  name: { type: String },
  cls: { type: String, default: "" },
});

// Auto-load icons
const icons = Object.fromEntries(
  Object.entries(import.meta.glob("~/assets/svg/*.svg", { as: "raw" })).map(
    ([key, value]) => {
      const filename = key.split("/").pop().split(".").shift();
      return [filename, value];
    }
  )
);

// Lazily load the icon
const icon = props.name && (await icons?.[props.name]?.());
</script>
```

For reference, in Nuxt 2 this component worked like this:

```
<template>
  <span class="svg-icon-ctr" :class="[name, cls]" v-html="require(`~/assets/svg/${name}.svg?raw`)"></span>
</template>

<script>
  export default {
    props: {
      name: { type: String },
      cls: { type: String, default: "" },
    },
    methods: {
      getClass() {
        return [ this.name, this.cls ].join(' ');
      }
    },
  };
</script>
```

**Comment by 3aluw** [comment]
> I solved it by using the following code:
> 
> ```
> <template>
>   <span class="svg-icon-ctr" :class="[name, cls]" v-html="icon"></span>
> </template>
> 
> <script setup>
> const props = defineProps({
>   name: { type: String },
>   cls: { type: String, default: "" },
> });
> 
> // Auto-load icons
> const icons = Object.fromEntries(
>   Object.entries(import.meta.glob("~/assets/svg/*.svg", { as: "raw" })).map(
>     ([key, value]) => {
>       const filename = key.split("/").pop().split(".").shift();
>       return [filename, value];
>     }
>   )
> );
> 
> // Lazily load the icon
> const icon = props.name && (await icons?.[props.name]?.());
> </script>
> ```
> 
> For reference, in Nuxt 2 this component worked like this:
> 
> ```
> <template>
>   <span class="svg-icon-ctr" :class="[name, cls]" v-html="require(`~/assets/svg/${name}.svg?raw`)"></span>
> </template>
> 
> <script>
>   export default {
>     props: {
>       name: { type: String },
>       cls: { type: String, default: "" },
>     }
>   };
> </script>
> ```

Sounds interesting. From a performance perspective, is it more advantageous to import all the files and load them lazily, or simply use the public folder?

**Comment by 3aluw** [comment]
> I solved it by using the following code:
> 
> ```
> <script setup>
> // Auto-load icons
> const icons = Object.fromEntries(
>   Object.entries(import.meta.glob("~/assets/svg/*.svg", { as: "raw" })).map(
>..........
>   )
> );
> 
> // Lazily load the icon
> const icon = props.name && (await icons?.[props.name]?.());
> </script>

> ```

Sounds interesting. 
From a performance perspective,
 is it more advantageous to import all the files and load them lazily, or simply use the public folder?

**Comment by matias-gmg** [comment]
> Sounds interesting. From a performance perspective, is it more advantageous to import all the files and load them lazily, or simply use the public folder?

I am not sure to be honest... Maybe someone from the team can clarify this?

**Comment by danielroe** [comment]
If your files don't change in content, it is more performant (and a better decision, I think) to add them to the public folder.

**Comment by bitbytebit1** [comment]
Finding it really frustrating to work with images, this needs to be properly addressed in the documentation: making it very clear that it is not possible to using images with dynamic names unless import.meta.glob is used.

**Comment by netopolit** [comment]
> Finding it really frustrating to work with images, this needs to be properly addressed in the documentation: making it very clear that it is not possible to using images with dynamic names unless import.meta.glob is used.

This is partially documented in the Vite guide, but I agree it should also be clarified in Nuxt docs.

**Comment by guidocaru** [comment]
can't believe this doesn't work

**Comment by TimGuendel** [comment]
Bump ☝️

**Comment by manniL** [comment]
I've written a [blog post about that topic recently](https://www.lichter.io/articles/nuxt3-vue3-dynamic-images/), but the easiest way would be using the `public` folder unless you have **a good reason not to**. In this case, [using the `import.meta.glob` solution from Daniel](https://github.com/nuxt/nuxt/issues/14766#issuecomment-1397365205) would be the best way. But loading assets that way is less performant than via `public` folder.

**Comment by shrpne** [comment]
It is very sad, that the best solution for using dynamic assets with Vite is not to use dynamic assets. After years with Webpack, optimizing images by the bundler seems like a basic task, but not with Vite

**Comment by shrpne** [comment]
@netopolit 
> I also believe this is a Vite limitation, and the issue should be raised in [vitejs/vite repository](https://github.com/vitejs/vite). Please correct me if I'm wrong.

Yes, but the Vite team closes such issues since this limitation is "by design". You should have significant weight in the community to be heard by Vite team

**Comment by maltseva-k** [comment]
Hi!
You can use src without '~/assets':
<img :src="`/${dynamic_image_name}`" alt="Discover Nuxt 3" />
It works for me :)
try it too

**Comment by BorisKamp** [comment]
I have a set of country flag icons in my `/public/country-flags` folder. When I set the SRC like this:
```
<img v-if="agent.attributes?.country_iso" :src="`/country-flags/${agent.attributes.country_iso}.svg`" class="h-6 w-6 me-2">
```

It does work when running local, however, when running on prod, the images return a 404 as the path is not converted:
`<img src="/country-flags/NL.svg" class="h-6 w-6 me-2">`

For images in the same `/public` folder that I set the SRC non-dynamic, the src is converted like so on prod and the image loads:
`<img class="h-8 w-auto" src="/logo-white.svg" alt="Todays logo">`

I have no clue how to fix this, I can't be the only one with this problem.

**Comment by joelee1992** [comment]
I assume there is no solution to this issue, even after attempting to use a new new URL(link, import.meta.url).href which only works locally but not in the production environment

**Comment by manniL** [comment]
@joelee1992 There is, as Daniel pointed out in https://github.com/nuxt/nuxt/issues/14766#issuecomment-1397365205 😊

**Comment by BorisKamp** [comment]
> @joelee1992 There is, as Daniel pointed out in [#14766 (comment)](https://github.com/nuxt/nuxt/issues/14766#issuecomment-1397365205) 😊

So the workaround is to move the files to the assets folder and use Daniels solution?

**Comment by manniL** [comment]
> So the workaround is to move the files to the assets folder and use Daniels solution?

@BorisKamp Well, you have two options (also [explained here](https://www.lichter.io/articles/nuxt3-vue3-dynamic-images/)):

* Use the public folder (recommended and easier)
* *Or* use the assets folder + use the solution daniel provided if necessary

**Comment by BorisKamp** [comment]
> > So the workaround is to move the files to the assets folder and use Daniels solution?
> 
> @BorisKamp Well, you have two options (also [explained here](https://www.lichter.io/articles/nuxt3-vue3-dynamic-images/)):
> 
> * Use the public folder (recommended and easier)
> * _Or_ use the assets folder + use the solution daniel provided if necessary

Thank you for your reply but that's the whole point: I am using the public folder, see my comment [up here](https://github.com/nuxt/nuxt/issues/14766#issuecomment-1844965126)

**Comment by manniL** [comment]
@BorisKamp Sorry, didn't see it was linked to the comment above. Do you have a minimal reproduction? Public folder should work in any case, no matter if referenced in the app or not 🤔

**Comment by BorisKamp** [comment]
> @BorisKamp Sorry, didn't see it was linked to the comment above. Do you have a minimal reproduction? Public folder should work in any case, no matter if referenced in the app or not 🤔

@manniL no problem man! Hmm do you know if I can reproduce a production environment locally? As my issue describes, this is only happening once deployed, with dynamic filename generation. I can verify that with static filename specification this does not happen.

**Comment by megaarmos** [comment]
I think i just found solution for this problem. I used ```<NuxtImg>``` for dynamic images. It worked both during development and in my production build
```
<NuxtImg :src="'images/' + yourImage" />
```

**Comment by goulashsoup** [comment]
@danielroe Is this something you consider to change in the future?

The presented "solutions" are actually all workarounds...

**Comment by danielroe** [comment]
I'm afraid this is just how Vite works. See https://github.com/nuxt/nuxt/issues/14766#issuecomment-1397365205.

This isn't a bug or something we can or should change in Nuxt, and this issue is simply open for us to add a note to the docs that points people to https://vitejs.dev/guide/features.html#glob-import with perhaps the example I provided above.

**Comment by GBKS** [comment]
I also ran into this issue, causing 10+ MB initial page loads, due to all the images being fetched. My current solution (which Github Copilot came up with) is to use the composable as described above, but with `eager: false`, which you can [see here](https://github.com/GBKS/nosta-me/pull/79/files#diff-32714c312b49ae14ceff36140dcab59157012ee70035d07ada1a129a807087d8). Now, `useAssets` needs to be asynchronous, requiring some changes on how images are used in components, but it really isn't too bad, although I wish this was generally easier to do. The site is much snappier now. Let me know if you see any problems with this approach (thanks in advance).

**Comment by spreadpando** [comment]
I would love to use vite and vue for all the synchronicity they offer together, however I will say that in this instance, and given the advanced bundling and build requirements of many of my vue apps, I've just opted to use the webpack-builder which allows me to use require or dynamic imports,  and really doesn't slow my dev process at all. 

I use a webpack loader to bundle, hash and replace image map dependency urls in .gltf files which can be dynamically imported across my vue components, are hashed at build time for cache-busting and archival purposes.

Having all the beneifts and feature-richness of webpack and the flexibility of Nuxt 3 has been an absolute dream and I'd highly recommend anyone assessing the framework, to consider trying the webpack-builder if your requirements neccesitate! In my use cases the massive gains in productivity, code quality and build versioning vastyle outweigh the few extra ms of build time.

Thank you to the team for continuing to support webpack. It really has made me confident in Nuxt 3's strategy and has allowed my small company to benefit enormously from both ecosystems without compromise.

**Comment by drewbaker** [comment]
I want to point out, that with a full dynamic image `src`, even using a `public` image it doesn't work.

This does not work:
```
const allLogos = ['/images/logo.png']

// Template
<img
    v-for="logo in allLogos"
    :key="logo"
    :src="logo"
>
```

This does work:
```
const allLogos = ['logo.png']

// Template
<img
    v-for="logo in allLogos"
    :key="logo"
    :src="`/images/${logo}`"
>
```

This was a huge GOTCHA for me. Hope it helps. I'm so sad at the state of modern frontend DX.

**Comment by danielroe** [comment]
@drewbaker  are you hosting with a custom baseUrl and want this to be automatically added to those urls?

**Comment by drewbaker** [comment]
> [@drewbaker](https://github.com/drewbaker) are you hosting with a custom baseUrl and want this to be automatically added to those urls?

I do not have a custom baseUrl set. I guess it's just the default. I am using Nuxt 4, so have my root has a `/app`. See screenshot for the directory structure. I thought this was default Nuxt 4 structure. 

![Image](https://github.com/user-attachments/assets/e5c55d00-fc5e-49b1-b789-75757e07d120)

.

**Comment by danielroe** [comment]
@drewbaker yes you should be good. would you open a new issue with a reproduction? that should absolutely work in both of the examples you listed.
